File: python/forth/forth.py
"""Forth interpretter."""
import operator
import re


NUM_RE = re.compile(r"^-?\d+$")


class StackUnderflowError(Exception):
    """Raised when the stack does not contain sufficient items for an operation."""


OPS = {
    "+": operator.add,
    "-": operator.sub,
    "*": operator.mul,
    "/": operator.floordiv,
    
}

# Minimum stack size for these operators.
# Anything smaller triggers a StackUnderflowError.
STACK_OPS = {
    0: [":"],
    1: ["dup", "drop"],
    2: ["+", "-", "*", "/", "swap", "over"],
}
MIN_STACK = {
    op: count for count, ops in STACK_OPS.items() for op in ops
}


class Stack(list):
    """Stack of values."""

    def dup(self):
        self.append(self[-1])

    def drop(self):
        self.pop()

    def swap(self):
        self.insert(-1, self.pop())

    def over(self):
        self.append(self[-2])

    @property
    def ops(self):
        ops = ["dup", "drop", "swap", "over"]
        return {op: getattr(self, op) for op in ops}


class Forth:
    """Forth interpretter."""

    def __init__(self):
        """Initialize."""
        self.stack: Stack[int] = Stack()
        self.defs: dict[str, list[str]] = {}

    def eval(self, data: list[str]) -> list[int]:
        """Evaluate a program, returning the stack."""
        for line in data:
            self.eval_chunk(line.lower().split())
        return self.stack

    def eval_chunk(self, words: list[str]) -> None:
        """Evaluate a chunk of instructions."""
        instructions = iter(words)
        while (word := next(instructions, None)) is not None:
            # Push numbers to the stack.
            if NUM_RE.match(word):
                self.stack.append(int(word))
                continue

            # Evaluate custom definitions.
            if word in self.defs:
                self.eval_chunk(self.defs[word])
                continue

            # Evaluate built-in operators.
            # Check the operator and stack size.
            if word not in MIN_STACK:
                raise ValueError("undefined operation")
            if len(self.stack) < MIN_STACK[word]:
                raise StackUnderflowError("Insufficient number of items in stack")

            self.alu_check(word)
            if word in "+-*/":
                # Perform the operation using the last two values on the stack.
                self.stack.append(OPS[word](self.stack.pop(-2), self.stack.pop(-1)))
            elif word in self.stack.ops:
                self.stack.ops[word]()
            elif word == ":":
                self.parse_definition(instructions)

    def alu_check(self, word):
        """Perform any arithmetic checks needed."""
        if word == "/" and self.stack[-1] == 0:
            raise ZeroDivisionError("divide by zero")

    def parse_definition(self, instructions):
        """Parse and set user defined values."""
        # Read definition name.
        name = next(instructions)
        # Check if the name is an int. Integers cannot be redefined.
        if NUM_RE.match(name):
            raise ValueError("illegal operation")
        # Read instructions until a ";" is found.
        vals = []
        while (cmd := next(instructions)) != ";":
            # Dereference custom definitions.
            if cmd in self.defs:
                vals.extend(self.defs[cmd])
            else:
                vals.append(cmd)
        self.defs[name] = vals


def evaluate(input_data: list[str]) -> list[int]:
    """Evaluate a Forth program."""
    return Forth().eval(input_data)

File: go/forth/forth.go
package forth

import (
	"fmt"
	"strconv"
	"strings"
)

// emulator contains emulator details: the stack and any custom definitions.
type emulator struct {
	stack     []int
	operators map[string][]string
}

// define handles a ": name ... ;" definition.
func (e *emulator) define(line []string) error {
	l := len(line)
	if l < 4 || line[0] != ":" || line[l-1] != ";" {
		return fmt.Errorf("malformed define, %v", line)
	}
	// The custom command name cannot be an integer.
	if _, err := strconv.Atoi(line[1]); err == nil {
		return fmt.Errorf("invalid define %s", line[1])
	}
	var operation []string
	for _, w := range line[2 : l-1] {
		// When storing a definition, expand existing definitions.
		if o, ok := e.operators[w]; ok {
			operation = append(operation, o...)
		} else {
			operation = append(operation, w)
		}
	}
	e.operators[line[1]] = operation
	return nil
}

// Evalute a command that operates on the stack.
func (e *emulator) stackOp(op string, count int) error {
	l := len(e.stack)
	if l < count {
		return fmt.Errorf("%s requires %d values; stack size is %d", op, count, l)
	}
	// Read the operands from the stack.
	var result, operands []int
	for i := range count {
		operands = append(operands, e.stack[l-count+i])
	}
	switch op {
	case "+":
		result = []int{operands[0] + operands[1]}
	case "-":
		result = []int{operands[0] - operands[1]}
	case "*":
		result = []int{operands[0] * operands[1]}
	case "/":
		if op == "/" && operands[1] == 0 {
			return fmt.Errorf("illegal division by zero")
		}
		result = []int{operands[0] / operands[1]}
	case "DUP":
		result = []int{operands[0], operands[0]}
	case "DROP":
		result = []int{}
	case "SWAP":
		result = []int{operands[1], operands[0]}
	case "OVER":
		result = []int{operands[0], operands[1], operands[0]}
	}
	// Update the stack.
	e.stack = append(e.stack[:l-count], result...)
	return nil
}

// eval evaluates one line.
func (e *emulator) eval(line []string) error {
	if line[0] == ":" {
		return e.define(line)
	}
	for _, word := range line {
		var err error
		var val int
		if val, err = strconv.Atoi(word); err == nil {
			// Integers get pushed to the stack.
			e.stack = append(e.stack, val)
		} else if v, ok := e.operators[word]; ok {
			// Custom definitions take precendence over built-in operators.
			err = e.eval(v)
		} else {
			switch word {
			case "+", "-", "*", "/":
				err = e.stackOp(word, 2)
			case "DUP", "DROP":
				err = e.stackOp(word, 1)
			case "SWAP", "OVER":
				err = e.stackOp(word, 2)
			default:
				return fmt.Errorf("invalid command %s", word)
			}
		}
		if err != nil {
			return err
		}
	}
	return nil
}

// Forth emulates a Forth program.
func Forth(input []string) ([]int, error) {
	e := &emulator{nil, map[string][]string{}}
	for _, line := range input {
		if err := e.eval(strings.Fields(strings.ToUpper(line))); err != nil {
			return nil, err
		}
	}
	return e.stack, nil
}
